File: library/src/main/java/com/ftinc/kit/mvp/BaseActivity.java
/*
 * Copyright (c) 2015 52inc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ftinc.kit.mvp;

import android.os.Bundle;
import android.support.v7.app.ActionBarActivity;
import android.support.v7.widget.Toolbar;

import com.ftinc.kit.R;

import butterknife.ButterKnife;
import dagger.ObjectGraph;

/**
 * This is a base UI activity that assists in creating a scoped
 * object graph on the activity for DI
 *
 * Project: Chipper
 * Package: com.r0adkll.chipper.ui
 * Created by drew.heavner on 11/12/14.
 */
public abstract class BaseActivity extends ActionBarActivity {

    protected Toolbar actionBarToolbar;
    private ObjectGraph activityGraph;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
//        activityGraph = HaystackApp.get(this).createScopedGraph(getModules());
        activityGraph.inject(this);
    }

    @Override
    public void setContentView(int layoutResID) {
        super.setContentView(layoutResID);
        getActionBarToolbar();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        activityGraph = null;
    }

    /**
     * Get the toolbar actionbar
     *
     * @return      get teh action bar
     */
    protected Toolbar getActionBarToolbar() {
        if (actionBarToolbar == null) {
            actionBarToolbar = ButterKnife.findById(this, R.id.toolbar_actionbar);
            if (actionBarToolbar != null) {
                setSupportActionBar(actionBarToolbar);
            }
        }
        return actionBarToolbar;
    }

    protected abstract Object[] getModules();
}

File: library-arch/src/main/java/com/ftinc/kit/arch/presentation/BaseActivity.kt
/*
 * Copyright (c) 2019 52inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.ftinc.kit.arch.presentation


import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import com.ftinc.kit.arch.R
import com.ftinc.kit.arch.presentation.delegates.ActivityDelegate
import com.ftinc.kit.arch.util.bindOptionalView
import io.reactivex.disposables.CompositeDisposable


abstract class BaseActivity : AppCompatActivity() {

    protected val appbar: Toolbar? by bindOptionalView(R.id.appbar)
    protected val disposables = CompositeDisposable()
    protected val delegates: ArrayList<ActivityDelegate> = ArrayList()


    protected abstract fun setupComponent()


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setupComponent()

        delegates.forEach { it.onCreate(savedInstanceState) }
    }


    override fun setContentView(layoutResID: Int) {
        super.setContentView(layoutResID)
        setSupportActionBar(appbar)
    }


    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        delegates.forEach { it.onSaveInstanceState(outState) }
    }


    override fun onResume() {
        super.onResume()
        delegates.forEach { it.onResume() }
    }


    override fun onStart() {
        super.onStart()
        delegates.forEach { it.onStart() }
    }


    override fun onPause() {
        super.onPause()
        delegates.forEach { it.onPause() }
    }


    override fun onStop() {
        super.onStop()
        delegates.forEach { it.onStop() }
    }


    override fun onDestroy() {
        disposables.clear()
        delegates.forEach { it.onDestroy() }
        super.onDestroy()
    }
}